swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
+        c: The number of cores along the coordinate.
+        halo_width: The halo width
+
+      Returns:
+        The local indices of the image points along a given coordinate.
+      """
+      index = tf.nest.map_structure(
+          lambda xi: tf.cast(xi // h, dtype=tf.int32) % (n - 2 * halo_width),
+          ip_coord,
+      )
+
+      ## index = tf.cond(
+      ##     all_coord[dim] == 0,
+      ##     lambda: tf.math.maximum(index, halo_width),
+      ##     lambda: index
+      ## )
+      ## index = tf.cond(
+      ##     all_coord[dim] == c - 1,
+      ##     lambda: tf.math.minimum(index, n - halo_width),
+      ##     lambda: index
+      ## )
+
+      return index
+
+    # Indices for the first image point layer (required for Dirichlet and
+    # Neumann boundary conditions)
+    p = get_local_index(
+        x_ip, params.dx, params.nx, 0, params.cx, params.halo_width
+    )
+    q = get_local_index(
+        y_ip, params.dy, params.ny, 1, params.cy, params.halo_width
+    )
+    s = get_local_index(
+        z_ip, params.dz, params.nz, 2, params.cz, params.halo_width
+    )
+
+    # Indices for the second image point layer (required for Neumann boundary
+    # conditions only)
+    p2 = get_local_index(
+        x_ip2, params.dx, params.nx, 0, params.cx, params.halo_width
+    )
+    q2 = get_local_index(
+        y_ip2, params.dy, params.ny, 1, params.cy, params.halo_width
+    )
+    s2 = get_local_index(
+        z_ip2, params.dz, params.nz, 2, params.cz, params.halo_width
+    )
+
+    def get_interp_weights(
+      image_point: tf.Tensor,
+      lower_left_index: tf.Tensor,
+    ) -> tf.Tensor:
+      """Computes the interpolation weights of each grid node enclosing a given
+      image point."""
+
+      node_dist = []
+      temp_mask = []
+
+      for i, j, k in itertools.product(range(2), range(2), range(2)):
+        p_x = common_ops.gather(xx,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+        p_y = common_ops.gather(yy,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+        p_z = common_ops.gather(zz,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+
+        node_dist.append(
+            tf.math.sqrt(
+                (image_point[0] - p_x) ** 2 +
+                (image_point[1] - p_y) ** 2 +
+                (image_point[2] - p_z) ** 2))
+
+        # Flag corner points with 1 which are in the fluid domain, and
+        # set all corner points to zero in the solid domain
+        temp_mask.append(
+            common_ops.gather(tf.transpose(ib_interior_mask, perm=[1, 2, 0]),
```

Comment:
`ib_interior_mask` contains halos if it's generated from a previous call to partial_mesh_for_core. So I believe what we should use here is ib_interior_mask[halo_width:-halo_width, halo_width:-halo_width, halo_width:-halo_width].